metrique/src/multi_flex.rs

```diff
@@ -0,0 +1,189 @@
+    /// Add an item to the MultiFlex collection.
+    ///
+    /// The item will be assigned the next available index and its metrics will be
+    /// prefixed accordingly when the collection is emitted.
+    ///
+    /// # Arguments
+    /// * `item` - The item to add to the collection
+    pub fn push(&mut self, item: T) {
+        self.0.push(item);
+    }
+
+    /// Creates a MulitFlex with a
+    pub fn with_capacity(cap: usize) -> Self {
+        Self(Vec::with_capacity(cap))
+    }
+}
+
+/// The closed form of a [`MultiFlex`] collection, containing the processed entries
+/// ready for metric emission.
+///
+/// This struct is created when a `MultiFlex<T>` is closed and contains each item
+/// in both unclosed (for prefix computation) and closed (for writing) forms.
+pub struct MultiFlexEntry<T, C> {
+    entries: Vec<(usize, T, C)>,
+}
+
+impl<T: CloseValueRef + FlexItem> CloseValue for MultiFlex<T> {
+    type Closed = MultiFlexEntry<T, T::Closed>;
+
+    fn close(self) -> Self::Closed {
+        MultiFlexEntry {
+            entries: self
+                .0
+                .into_iter()
+                .enumerate()
+                .map(|(idx, item)| {
+                    let closed = item.close_ref();
+                    (idx, item, closed)
+                })
+                .collect(),
+        }
+    }
+}
+
+impl<T: CloseValueRef + FlexItem + Clone> CloseValue for &MultiFlex<T> {
+    type Closed = MultiFlexEntry<T, T::Closed>;
+
+    fn close(self) -> Self::Closed {
+        MultiFlexEntry {
+            entries: self
+                .0
+                .iter()
+                .enumerate()
+                .map(|(idx, item)| {
+                    let closed = item.close_ref();
+                    (idx, item.clone(), closed)
+                })
+                .collect(),
+        }
+    }
+}
+
+struct EmptyString;
+impl ConstStr for EmptyString {
+    const VAL: &'static str = "";
+}
+
+impl<T: FlexItem, C: InflectableEntry<NoPrefix<NS>>, NS: NameStyle> InflectableEntry<NS>
+    for MultiFlexEntry<T, C>
+{
+    fn write<'a>(&'a self, writer: &mut impl EntryWriter<'a>) {
+        let external_prefix =
+            const_str_value::<NS::Inflect<EmptyString, EmptyString, EmptyString, EmptyString>>();
+        // Pre-allocate a string buffer for building prefixes
+        let mut entry = SmartString::<LazyCompact>::new();
+
+        for (idx, item, closed) in &self.entries {
+            entry.truncate(0);
+            item.prefix_item(*idx, &mut entry);
+
+            let mut prefixer = DynPrefix {
+                head: external_prefix.as_ref(),
+                prefix: entry.as_str(),
+                writer,
+            };
+            InflectableEntry::<NoPrefix<NS>>::write(closed, &mut prefixer);
+        }
+    }
+
+    fn sample_group(&self) -> impl Iterator<Item = SampleGroupElement> {
+        // sample groups are ignored inside of multi-groups
+        vec![].into_iter()
+    }
+}
+
+struct DynPrefix<'b, T> {
```

Comment:
(you can hold `idx` and `item` within the `DynPrefix` and then you will not need the `SmartString`. You'll still need to allocate a `String` for every item.